internal/pkg/docker/scheduler/scheduler.go

```diff
@@ -0,0 +1,312 @@
+
+package scheduler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"maps"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/aws/copilot-cli/internal/pkg/docker/dockerengine"
+)
+
+// Scheduler manages running a Task. Only a single Task
+// can be running at a time for a given Scheduler. A Scheduler
+// can only be Start()-ed once; multiple calls to Start() will
+// break things.
+type Scheduler struct {
+	idPrefix   string
+	logOptions logOptionsFunc
+
+	curTask   Task
+	curTaskID atomic.Int32
+	runErrs   chan error
+	stopped   chan struct{}
+	wg        *sync.WaitGroup
+	actions   chan action
+	stopOnce  *sync.Once
+
+	docker DockerEngine
+}
+
+type action interface {
+	Do(s *Scheduler) error
+}
+
+type logOptionsFunc func(name string, ctr ContainerDefinition) dockerengine.RunLogOptions
+
+// DockerEngine is used by Scheduler to manage containers.
+type DockerEngine interface {
+	Run(context.Context, *dockerengine.RunOptions) error
+	IsContainerRunning(context.Context, string) (bool, error)
+	Stop(context.Context, string) error
+}
+
+// NewScheduler creates a new Scheduler. idPrefix is a prefix used when
+// naming containers that are run by the Scheduler.
+func NewScheduler(docker DockerEngine, idPrefix string, logOptions logOptionsFunc) *Scheduler {
+	s := &Scheduler{
+		idPrefix:   idPrefix,
+		logOptions: logOptions,
+		stopped:    make(chan struct{}),
+		docker:     docker,
+		wg:         &sync.WaitGroup{},
+		stopOnce:   &sync.Once{},
+		actions:    make(chan action),
+		runErrs:    make(chan error),
+	}
+	return s
+}
+
+// Start starts the Scheduler. Start must be called before any other
+// Scheduler functions. Errors from containers run by the scheduler
+// or from Scheduler actions are sent to the returned error channel.
+// The returned error channel is closed after calling Stop() has
+// stopped the Scheduler.
+func (s *Scheduler) Start() chan error {
+	// close done when all goroutines created by scheduler have finished
+	done := make(chan struct{})
+	errs := make(chan error)
+
+	// scheduler routine
+	s.wg.Add(1) // decremented by stopAction
+	go func() {
+		for {
+			select {
+			case action := <-s.actions:
+				if err := action.Do(s); err != nil {
+					errs <- err
+				}
+			case err := <-s.runErrs:
+				errs <- err
+			case <-done:
+				close(errs)
+				return
+			}
+		}
+	}()
+
+	go func() {
+		s.wg.Wait()
+		close(done)
+	}()
+
+	return errs
+}
+
+// RunTask stops the current running task and starts task.
+func (s *Scheduler) RunTask(task Task) {
+	// this guarantees the following:
+	// - if runTaskAction{} is pulled by the scheduler, any errors
+	//   returned by it are reported by the scheduler.
+	// - if Stop() is called _before_ the scheduler picks up this
+	//   action, then this action is skipped.
+	select {
+	case <-s.stopped:
+	case s.actions <- &runTaskAction{
+		task: task,
+	}:
+	}
+}
+
+type runTaskAction struct {
+	task Task
+}
+
+func (r *runTaskAction) Do(s *Scheduler) error {
```

Comment:
I feel like `runTask` might not accurately describe what this function does: it is also stopping the old task if it exists. Or maybe we need a more specific name for the scheduler since it's not general purpose for now 🤔 Also more explanation on how "pause" container work in scheduler comment 💭

Reply:
based on some offline discussion, i renamed `Scheduler` to `Orchestrator`! Let me know if you think `RunTask()` still doesn't make sense with that naming context.